File: go/unionfind/union_find4.go
package unionfind

import (
	"errors"
)

// 基于 rank 的优化
type UnionFind4 struct {
	parent []int
	rank 	 []int
}

func NewRankUnionFind(size int) *UnionFind4 {
	parent := make([]int, size)
	rank := make([]int, size)

	for i := 0; i < size; i++ {
		parent[i] = i
		rank[i] = 1
	}

	return &UnionFind4{parent, rank}
}

func (u *UnionFind4) GetSize() int {
	return len(u.parent)
}

func (u *UnionFind4) find(p int) (int, error) {
	if p < 0 || p >= len(u.parent) {
		return 0, errors.New("Index out of range")
	}

	for ; p != u.parent[p]; p = u.parent[p] {}
	return p, nil
}

func (u *UnionFind4) IsConnected(p, q int) bool {
	pPar, err := u.find(p)
	if err != nil {
		return false
	}
	qPar, err := u.find(q)
	if err != nil {
		return false
	}

	return qPar == pPar
}

func (u *UnionFind4) UnionElements(p, q int) error {
	pPar, err := u.find(p)
	if err != nil {
		return err
	}
	qPar, err := u.find(q)
	if err != nil {
		return err
	}

	if pPar == qPar {
		return nil
	}

	pRank := u.rank[pPar]
	qRank := u.rank[qPar]
	if pRank < qRank {
		u.parent[pPar] = qPar
	} else if pRank > qRank {
		u.parent[qPar] = pPar
	} else {
		// 这里哪边往哪边上并其实无所谓了
		u.parent[pPar] = qPar
		u.rank[qPar]++
	}
	return nil
}

File: js2/unionFind4.ts
import type { UnionFind } from './unionFind';

// quick union 版本 基于 rand 优化
export default class UnionFind4 implements UnionFind {
    private parent: Array<number>;
    private rank: Array<number>; 

    constructor(size: number) {
        this.parent = new Array(size);
        this.rank = new Array(size);

        for (let i = 0; i < size; i++) {
            this.parent[i] = i;
            this.rank[i] = 1;
        }
    }

    getSize() {
        return this.parent.length;
    }

    private find(p: number): number {
        while (this.parent[p] !== p) {
            p = this.parent[p];
        }
        return p;
    }

    // 路径压缩
    private find2(p: number): number {
        while (this.parent[p] !== p) {
            this.parent[p] = this.parent[this.parent[p]];
            p = this.parent[p];
        }
        return p;
    }

    // O(h)
    isConnected(p: number, q: number): boolean {
        return this.find(p) === this.find(q);
    }

    // O(h)
    unionElements(p: number, q: number) {
        const pParent = this.find(p);
        const qParent = this.find(q);
        if (pParent === qParent) {
            return;
        }
        if (this.rank[pParent] < this.rank[qParent]) {
            this.parent[pParent] = qParent;
        } else if (this.rank[pParent] > this.rank[qParent]) {
            this.parent[qParent] = pParent;
        } else {
            this.parent[pParent] = qParent;
            this.rank[qParent]++;
        }
    }
}